File: Lab7/С++/С++/LinkedList.cpp
#include <iostream>
#include "LinkedList.hpp"

LinkedList::LinkedList(){
    head = NULL;
    _size = 0;
}

void LinkedList::Push(int value){
    if (head == NULL){
        head = new Node(value, NULL);
        current = head;
        _size++;
        return;
    }
    auto newNode = new Node(value, NULL);
    current->next = newNode;
    current = newNode;
    _size++;
}


int LinkedList::DividedByFive(){
    int counter = 0;
    current = head;
    while (current != NULL) {
        if (current->value % 5 == 0){
            counter++;
        }
        current = current->next;
    }
    return counter;
}

int& LinkedList::operator[](const int index){
    int counter = 0;
    int result = 0;
    current = head;
    while (current != NULL){
        if (counter == index){
            result = current->value;
            break;
        }
        current = current->next;
        counter++;
    }
    return result = -1;
}


int LinkedList::Size() {
    return this->_size;
    
}

void LinkedList::RemoveAfterMax(){
    current = head;
    int maxElement = head->value;
    while (current != NULL){
        if (current->value > maxElement){
            maxElement = current->value;
        }
        current = current->next;
    }
    current = head;
    _size = 0;
    while (current != NULL){
        if (current->value == maxElement){
            current->next = NULL;
        }
        current = current->next;
        _size++;
    }
}

void LinkedList::PrintList(){
    current = head;
    while (current != NULL){
        std::cout << current->value << " -> ";
        current = current->next;
    }
    std::cout << "null" << std::endl;
}

File: Lab7/CSharp/LinkedList.cs
using System;
using System.Data;

namespace CSharp
{
    public class LinkedList
    {
        private Node _head;
        private Node _current;
        public int Size { get; private set; }

        public LinkedList()
        {
            _head = null;
            Size = 0;
        }

        public void Push(int value)
        {
            if (_head == null)
            {
                _head = new Node(value);
                _current = _head;
                Size++;
                return;
            }
            var newNode = new Node(value);
            _current.Next = newNode;
            _current = newNode;
            Size++;
        }
        
        public int DividedByFive()
        {
            var counter = 0;
            _current = _head;
            while (_current != null)
            {
                if (_current.Data % 5 == 0)
                {
                    counter++;
                }
                _current = _current.Next;
            }

            return counter;
        }

        public void RemoveAfterMax()
        {
            
            _current = _head;
            var maxElement = _head.Data;
            while (_current != null)
            {
                if (_current.Data > maxElement)
                {
                    maxElement = _current.Data;
                }

                _current = _current.Next;
            }


            _current = _head;
            Size = 0;
            while (_current != null)
            {
                if (_current.Data == maxElement)
                {
                    _current.Next = null;
                }
                _current = _current.Next;
                Size++;
            }
        }

        public int? this[int index]
        {
            get
            {
                if (index < 0 || index >= Size)
                {
                    Console.WriteLine("Out of range.");
                    return null;
                }

                var i = 0;
                _current = _head;
                while (i != index)
                {
                    _current = _current.Next;
                    i++;
                }
                
                return _current.Data;
            }
        }

        public void PrintLinkedList()
        {
            _current = _head;
            while (_current != null)
            {
                Console.Write(_current.Data + " -> ");
                _current = _current.Next;
            }

            Console.WriteLine( "null");
        }
        
    }
}